File: backend/src/utils/fileLock.ts
// backend/src/utils/fileLock.ts
import fs from 'fs';
import path from 'path';
import { logger } from './logger';

/**
 * File Locking Utility for Config Files
 * 
 * Provides a robust file-based locking mechanism to prevent race conditions
 * when multiple processes attempt to write to shared configuration files.
 * 
 * This prevents catastrophic config corruption that could bring down the entire stack.
 */

export class FileLockError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'FileLockError';
  }
}

export class FileLockTimeout extends FileLockError {
  constructor(message: string) {
    super(message);
    this.name = 'FileLockTimeout';
  }
}

interface LockInfo {
  pid: number;
  time: number;
  age_seconds?: number;
  error?: string;
}

/**
 * Acquire an exclusive lock on a config file.
 * 
 * Creates a .lock file to prevent concurrent writes. Uses try...finally
 * to guarantee the lock is released even if an exception occurs.
 * 
 * @param filename - Path to the config file to lock
 * @param timeout - Maximum time to wait for lock acquisition (milliseconds)
 * @param pollInterval - Time between lock acquisition attempts (milliseconds)
 * @returns Promise that resolves when lock is acquired
 * @throws FileLockTimeout if lock cannot be acquired within timeout period
 */
async function acquireLock(
  filename: string,
  timeout: number = 10000,
  pollInterval: number = 100
): Promise<() => void> {
  const lockFile = `${filename}.lock`;
  const startTime = Date.now();

  while (true) {
    try {
      // Try to create lock file exclusively (fails if exists)
      // Using wx flag ensures atomic creation
      const lockInfo = {
        pid: process.pid,
        time: Date.now(),
      };

      fs.writeFileSync(lockFile, JSON.stringify(lockInfo, null, 2), {
        flag: 'wx', // Create file, fail if exists
        mode: 0o644,
      });

      logger.debug(`Acquired lock on ${filename}`);

      // Return release function
      return () => {
        try {
          fs.unlinkSync(lockFile);
          logger.debug(`Released lock on ${filename}`);
        } catch (error) {
          if ((error as NodeJS.ErrnoException).code !== 'ENOENT') {
            logger.error(`Error releasing lock on ${filename}:`, error);
          }
        }
      };
    } catch (error) {
      const err = error as NodeJS.ErrnoException;

      // If file exists, wait and retry
      if (err.code === 'EEXIST') {
        const elapsed = Date.now() - startTime;

        if (elapsed >= timeout) {
          // Try to read lock file to see who has it
          let lockHolder = 'unknown';
          try {
            const lockContent = fs.readFileSync(lockFile, 'utf8');
            lockHolder = lockContent.trim();
          } catch {
            // Ignore read errors
          }

          throw new FileLockTimeout(
            `Failed to acquire lock on ${filename} after ${timeout}ms. Lock held by: ${lockHolder}`
          );
        }

        // Wait and retry
        logger.debug(`Waiting for lock on ${filename} (${elapsed}ms elapsed)`);
        await new Promise((resolve) => setTimeout(resolve, pollInterval));
      } else {
        // Other error, rethrow
        throw new FileLockError(`Failed to acquire lock on ${filename}: ${err.message}`);
      }
    }
  }
}

/**
 * Execute a function with an exclusive lock on a config file.
 * 
 * @param filename - Path to the config file to lock
 * @param fn - Function to execute while holding the lock
 * @param timeout - Maximum time to wait for lock acquisition (milliseconds)
 * @returns Promise that resolves with the function's return value
 * 
 * @example
 * await withConfigLock('/app/etc/pools.yaml', async () => {
 *   fs.writeFileSync('/app/etc/pools.yaml', yaml.dump(config));
 * });
 */
export async function withConfigLock<T>(
  filename: string,
  fn: () => T | Promise<T>,
  timeout: number = 10000
): Promise<T> {
  const releaseLock = await acquireLock(filename, timeout);

  try {
    return await fn();
  } finally {
    releaseLock();
  }
}

/**
 * Check if a file is currently locked.
 * 
 * @param filename - Path to the config file
 * @returns Lock info if locked, null if not locked
 */
export function checkLockStatus(filename: string): LockInfo | null {
  const lockFile = `${filename}.lock`;

  try {
    if (!fs.existsSync(lockFile)) {
      return null;
    }

    const content = fs.readFileSync(lockFile, 'utf8');
    const info = JSON.parse(content) as LockInfo;

    // Add lock file age
    const stats = fs.statSync(lockFile);
    info.age_seconds = (Date.now() - stats.mtimeMs) / 1000;

    return info;
  } catch (error) {
    logger.error(`Error reading lock file ${lockFile}:`, error);
    return {
      pid: 0,
      time: 0,
      error: error instanceof Error ? error.message : 'Unknown error',
    };
  }
}

/**
 * Force release a lock (use with caution!).
 * 
 * This should only be used in emergency situations where a lock
 * is stuck due to a crashed process.
 * 
 * @param filename - Path to the config file
 * @returns True if lock was released, false if no lock existed
 */
export function forceReleaseLock(filename: string): boolean {
  const lockFile = `${filename}.lock`;

  try {
    fs.unlinkSync(lockFile);
    logger.warn(`Force released lock on ${filename}`);
    return true;
  } catch (error) {
    const err = error as NodeJS.ErrnoException;
    if (err.code === 'ENOENT') {
      return false;
    }
    logger.error(`Error force releasing lock on ${filename}:`, error);
    throw error;
  }
}

/**
 * Convenience function for safely writing YAML config files with locking.
 * 
 * @example
 * await safeWriteYaml('/app/etc/pools.yaml', async () => {
 *   fs.writeFileSync('/app/etc/pools.yaml', yaml.dump(config));
 * });
 */
export async function safeWriteYaml<T>(
  filename: string,
  fn: () => T | Promise<T>,
  timeout: number = 10000
): Promise<T> {
  return withConfigLock(filename, fn, timeout);
}

/**
 * Convenience function for safely writing JSON config files with locking.
 * 
 * @example
 * await safeWriteJson('/app/etc/config.json', async () => {
 *   fs.writeFileSync('/app/etc/config.json', JSON.stringify(config));
 * });
 */
export async function safeWriteJson<T>(
  filename: string,
  fn: () => T | Promise<T>,
  timeout: number = 10000
): Promise<T> {
  return withConfigLock(filename, fn, timeout);
}

File: python-scheduler/file_lock.py
"""
File Locking Utility for Config Files

Provides a robust file-based locking mechanism to prevent race conditions
when multiple processes attempt to write to shared configuration files.

This prevents catastrophic config corruption that could bring down the entire stack.
"""

import os
import time
import logging
from contextlib import contextmanager
from typing import Generator, Optional
from pathlib import Path

logger = logging.getLogger(__name__)


class FileLockError(Exception):
    """Raised when a file lock cannot be acquired"""
    pass


class FileLockTimeout(FileLockError):
    """Raised when lock acquisition times out"""
    pass


@contextmanager
def with_config_lock(
    filename: str,
    timeout: float = 10.0,
    poll_interval: float = 0.1
) -> Generator[None, None, None]:
    """
    Context manager that acquires an exclusive lock on a config file.
    
    Creates a .lock file to prevent concurrent writes. Uses a try...finally
    block to guarantee the lock is released even if an exception occurs.
    
    Args:
        filename: Path to the config file to lock
        timeout: Maximum time to wait for lock acquisition (seconds)
        poll_interval: Time between lock acquisition attempts (seconds)
    
    Raises:
        FileLockTimeout: If lock cannot be acquired within timeout period
        
    Example:
        with with_config_lock('/app/etc/pools.yaml'):
            # Write to file safely
            with open('/app/etc/pools.yaml', 'w') as f:
                yaml.dump(config, f)
    """
    lock_file = f"{filename}.lock"
    lock_acquired = False
    start_time = time.time()
    
    try:
        # Try to acquire lock
        while True:
            try:
                # Create lock file exclusively (fails if exists)
                # Using os.O_CREAT | os.O_EXCL ensures atomic creation
                fd = os.open(lock_file, os.O_CREAT | os.O_EXCL | os.O_WRONLY, 0o644)
                
                # Write process info to lock file for debugging
                lock_info = f"PID: {os.getpid()}\nTime: {time.time()}\n"
                os.write(fd, lock_info.encode())
                os.close(fd)
                
                lock_acquired = True
                logger.debug(f"Acquired lock on {filename}")
                break
                
            except FileExistsError:
                # Lock file exists, check if it's stale
                elapsed = time.time() - start_time
                
                if elapsed >= timeout:
                    # Try to read lock file to see who has it
                    lock_holder = "unknown"
                    try:
                        with open(lock_file, 'r') as f:
                            lock_holder = f.read().strip()
                    except Exception:
                        pass
                    
                    raise FileLockTimeout(
                        f"Failed to acquire lock on {filename} after {timeout}s. "
                        f"Lock held by: {lock_holder}"
                    )
                
                # Wait and retry
                logger.debug(f"Waiting for lock on {filename} ({elapsed:.1f}s elapsed)")
                time.sleep(poll_interval)
        
        # Lock acquired, yield control to caller
        yield
        
    finally:
        # Always release lock, even if exception occurred
        if lock_acquired:
            try:
                os.remove(lock_file)
                logger.debug(f"Released lock on {filename}")
            except FileNotFoundError:
                # Lock file already removed (shouldn't happen, but not critical)
                logger.warning(f"Lock file {lock_file} was already removed")
            except Exception as e:
                # Log but don't raise - we don't want to mask the original exception
                logger.error(f"Error releasing lock on {filename}: {e}")


def check_lock_status(filename: str) -> Optional[dict]:
    """
    Check if a file is currently locked.
    
    Args:
        filename: Path to the config file
        
    Returns:
        Dict with lock info if locked, None if not locked
    """
    lock_file = f"{filename}.lock"
    
    if not os.path.exists(lock_file):
        return None
    
    try:
        with open(lock_file, 'r') as f:
            content = f.read()
        
        # Parse lock info
        info = {}
        for line in content.strip().split('\n'):
            if ':' in line:
                key, value = line.split(':', 1)
                info[key.strip()] = value.strip()
        
        # Add lock file age
        lock_age = time.time() - os.path.getmtime(lock_file)
        info['age_seconds'] = lock_age
        
        return info
        
    except Exception as e:
        logger.error(f"Error reading lock file {lock_file}: {e}")
        return {'error': str(e)}


def force_release_lock(filename: str) -> bool:
    """
    Force release a lock (use with caution!).
    
    This should only be used in emergency situations where a lock
    is stuck due to a crashed process.
    
    Args:
        filename: Path to the config file
        
    Returns:
        True if lock was released, False if no lock existed
    """
    lock_file = f"{filename}.lock"
    
    try:
        os.remove(lock_file)
        logger.warning(f"Force released lock on {filename}")
        return True
    except FileNotFoundError:
        return False
    except Exception as e:
        logger.error(f"Error force releasing lock on {filename}: {e}")
        raise


# Convenience function for common use case
@contextmanager
def safe_write_yaml(filename: str, timeout: float = 10.0) -> Generator[None, None, None]:
    """
    Context manager for safely writing YAML config files with locking.
    
    Example:
        with safe_write_yaml('/app/etc/pools.yaml'):
            with open('/app/etc/pools.yaml', 'w') as f:
                yaml.dump(config, f)
    """
    with with_config_lock(filename, timeout=timeout):
        yield


@contextmanager
def safe_write_json(filename: str, timeout: float = 10.0) -> Generator[None, None, None]:
    """
    Context manager for safely writing JSON config files with locking.
    
    Example:
        with safe_write_json('/app/etc/config.json'):
            with open('/app/etc/config.json', 'w') as f:
                json.dump(config, f)
    """
    with with_config_lock(filename, timeout=timeout):
        yield
